### PR TITLE
J2735 2024 BSM Pipeline Improvements

### PR DESCRIPTION
**Pull request by drewjj**
# PR Details
## Description

Includes a complete overhaul of the BSM pipeline from both UDP and log offload endpoints.
- New J2735 2024 POJOs from the jpo-asn-pojos project have been incorporated into the BSM pipeline and all output now conforms 100% to the J2735 2024 as defined by the ASN.
- Unused and redundant POJO Kafka topics and classes have been removed. The Asn1DecoderRouter now handles BSM messages like the other message types and writes directly to the topic.OdeBsmJson topic.
- The log file offload and UDP endpoint BSM metadata now retains its rxSource value and is not unanimously overwritten with "NA".
- Unit tests have been updated to conform to the new data formats and pipeline.

## Related Issue

N/A

## How Has This Been Tested?

This has been tested locally in a single instance environment through the following methods:
- BSM UDP port 46800 endpoint using the BSM UDP send script found on the jpo-ode project's [scripts/tests directory](https://github.com/CDOT-CV/jpo-ode/tree/master/scripts/tests).
-  Log files dropped into the mounted `uploads/bsmlog` directory using the binary files found within the project's [data directory](https://github.com/CDOT-CV/jpo-ode/tree/master/data). All log file types and message type content were verified to still be functional.
- Unit tests were utilized to ensure known corner cases were tested properly.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/develop/docs/contributing_guide.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/decoder-output-bsm.xml [hunk 1]
```diff
@@ -1,237 +1,211 @@
...
-                  </SupplementalVehicleExtensions>
-                </partII-Value>
-              </BSMpartIIExtension>
-            </partII>
-          </BasicSafetyMessage>
-        </value>
-      </MessageFrame>
-    </data>
-  </payload>
+	<metadata>
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/decoder-output-bsm.xml:3]: Test data is going to be ignored for checkstyle comments since it is not necessarily intended to be read like the code would be.

[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/decoder-output-bsm.xml]: I addressed the tab to spaces thing anyways. Seems to be the only checkstyle complaint here.

**Comment by drewjj** [comment]
I still need to add documentation updates for pipeline structural changes.

There may be more unused POJOs that could be removed. Due to many of the BSM data structures also being used by other message types, the overall removal of unused POJOs will be a task done after all pipelines have been migrated to use the new POJOs. We can migrate the used ones to a new home as well if that is desirable.

File: jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/decoder-output-bsm.xml [hunk 2]
```diff
@@ -1,7 +1,6 @@
 <?xml version="1.0"?>
 <OdeAsn1Data>
   <metadata>
-    <bsmSource>EV</bsmSource>
```
Comment (mcook42) [inline jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/decoder-output-bsm.xml:4]: question(blocking): I understand why this line was deleted and why `<asn1>` changed, but I'm not sure why the rest of the data changed. I'm somewhat concerned about the deleted content at the end of the file. Can you talk about why so much of this test file changed?

File: jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/expected-bsm.json
Comment (mcook42) [inline jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/expected-bsm.json:1]: chore(blocking): I dropped this into an online JSON schema validator, and it looks like we'll need to update [schema-bsm.json](https://github.com/CDOT-CV/jpo-ode/blob/2e8b4d1a53c62e15788981d26d2e1f3a60c8d8e2/jpo-ode-core/src/main/resources/schemas/schema-bsm.json#L6) to align with this new JSON output

image

**Comment by drewjj** [comment]
> Overall, the changes look sound. Great work! I do have some blockers, though.
> 
> Question (non-blocking): With the removal of the OdeBsm Listener, it looks like there are no more usages of [MessagingDeserializer ](https://github.com/CDOT-CV/jpo-ode/blob/cba97fb8ffe9494a5f3bc6aaf7ea8db6163f1a73/jpo-ode-core/src/main/java/us/dot/its/jpo/ode/wrapper/serdes/MessagingDeserializer.java#L35)or [MessagingSerializer](https://github.com/CDOT-CV/jpo-ode/blob/cba97fb8ffe9494a5f3bc6aaf7ea8db6163f1a73/jpo-ode-core/src/main/java/us/dot/its/jpo/ode/wrapper/serdes/MessagingSerializer.java#L34). Should we keep these or delete them?

You seem to be right. I have removed these two classes and their associated tests.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/BsmLogFileParser.java [hunk 3]
```diff
@@ -34,7 +33,14 @@ public class BsmLogFileParser extends LogFileParser {
 
   private static final int DIRECTION_LENGTH = 1;
 
-  private BsmSource bsmSource; // 0 for EV(Tx), 1 for RV(Rx)
+  /**
+   * Enum representing the source of the BSM: 0 (EV/Tx), 1 (RV/Rx), or unknown.
+   */
+  public enum BsmSource {
+      EV, RV, unknown
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/BsmLogFileParser.java]: Addressed manually due to this sort of changing affecting more than just the line itself. But addressed.

File: jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/decoder-output-bsm.xml (hunk 2 above)
Comment (drewjj) [reply jpo-ode-svcs/src/test/resources/us/dot/its/jpo/ode/services/asn1/decoder-output-bsm.xml:4]: Addressed! Yes, I had used a new payload since it required some additional leg work to generate the XML output for the same payload. You should be very pleased with the change differences now.

**Comment by drewjj** [comment]
Waiting for a schema and then this pull request will be unmarked as a draft.

File: jpo-ode-core/src/main/resources/schemas/schema-bsm.json [hunk 4]
```diff
@@ -234,7 +230,7 @@
                 }
             },
             "required": [
-                "bsmSource",
+                "source",
                 "logFileName",
```
[the remark this answers is not in the document]
Comment (drewjj) [reply jpo-ode-core/src/main/resources/schemas/schema-bsm.json:234]: Interesting. I suppose the tests don't necessarily involve every step of our pipeline for each test. I am pretty confident in all actual use cases of the jpo-ode whether the UDP endpoint is used or the log file ingestion is used these fields would be included.

**Review by dmccoystephenson — APPROVED** [review]
Looks good!

- The unit tests pass
- The program starts up correctly
- BSMs are properly ingested by the UDP endpoint
- JSON in `topic.OdeBsmJson` passes the BSM schema

**Review by Michael7371 — COMMENTED** [review]
This code looks good, I'll wait to approve until the [schema PR](https://github.com/CDOT-CV/jpo-asn-pojos/pull/28) is complete and the schema URL reference is updated.
